discoveryserver/handlers/token.go

```diff
@@ -0,0 +1,119 @@
+package handlers
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"log"
+	"net/http"
+	"path"
+	"strconv"
+	"strings"
+	"time"
+
+	"go.etcd.io/etcd/discoveryserver/handlers/httperror"
+	"go.etcd.io/etcd/etcdserver/api/v2store"
+
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+func init() {
+	tokenCounter = prometheus.NewCounterVec(
+		prometheus.CounterOpts{
+			Name: "endpoint_token_requests_total",
+			Help: "How many /token requests processed, partitioned by status code and HTTP method.",
+		},
+		[]string{"code", "method"},
+	)
+	prometheus.MustRegister(tokenCounter)
+}
+
+var tokenCounter *prometheus.CounterVec
+
+func TokenHandler(ctx context.Context, w http.ResponseWriter, r *http.Request) {
+	st := ctx.Value(stateKey).(*State)
+
+	var ev *v2store.Event
+	var err error
+	switch r.Method {
+	case http.MethodGet:
+		var recursive bool
+		rq := r.URL.Query().Get("recursive")
+		recursive, err := strconv.ParseBool(rq)
+		if err != nil {
+			recursive = false
+		}
+
+		pq := r.URL.Query().Get("waitIndex")
+		waitIndex, err := strconv.ParseUint(pq, 10, 64)
+		if err != nil {
+			waitIndex = 0
+		}
+
+		wq := r.URL.Query().Get("wait")
+		if wq != "" {
+			watcher, err := st.v2.Watch(r.URL.Path, recursive, false, uint64(waitIndex+1))
+			go func() {
+				time.Sleep(time.Minute)
+				watcher.Remove()
+			}()
+			ev = <-watcher.EventChan()
+			watcher.Remove()
+			if err != nil || ev == nil {
+				log.Printf("TokenHandler: %v", err)
+				w.WriteHeader(http.StatusOK)
```

Comment:
this can be an error case (or a timeout?). should we just write statusOK?

Reply:
gah, great catch, done.